### PR TITLE
fix: Show RTL text correctly in Skia

### LINKED ISSUE #6783: [SKIA] RTL Language support on WPF SKIA (Persian, Arabic etc.)
## Current behavior and Expected behavior

In UWP when we type something in an RTL language, it shows as expected, with correct typing and flow direction and everything. But in SKIA head it does not render the text as expected. It doesn't follow the correct flow direction and it renders characters discreetly and not in a correct form. Please investigate the issue and fix it.

Screenshot 2021-08-16 183830

Right: The correct form in UWP
Left: The wrecked output by SKIA

## How to reproduce it 

```xaml
 <StackPanel Background="{ThemeResource ApplicationPageBackgroundThemeBrush}" FlowDirection="RightToLeft">
		<TextBlock Text="سلام به همه!" Margin="20" FontSize="30" />
        <TextBlock Margin="20" TextWrapping="Wrap" >
            زبان پارسی یک زبان چپ به راست است اما همه آن را با عربی اشتباه می گیرند!
        </TextBlock>
    </StackPanel>
```

## Works on UWP/WinUI
Yes

## Environment

- Uno.UI / Uno.UI.WebAssembly / Uno.UI.Skia

Nuget Package Version(s):

Affected platform(s):

  -  WPF

IDE:

- [ ] Visual Studio 2019 (Latest)

### PR DESCRIPTION
**Pull request by Youssef1313**
GitHub Issue (If applicable): closes #6783

## PR Type

What kind of change does this PR introduce?

- Bugfix

## What is the current behavior?

RTL languages doesn't render correctly.

## What is the new behavior?

They render correctly now.

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Docs have been added/updated which fit [documentation template](https://github.com/unoplatform/uno/blob/master/doc/.feature-template.md) (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](https://github.com/unoplatform/uno/blob/master/doc/articles/uno-development/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] Validated PR `Screenshots Compare Test Run` results.
- [ ] Contains **NO** breaking changes
- [ ] Associated with an issue (GitHub or internal) and uses the [automatic close keywords](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue).
- [ ] Commits must be following the [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/#summary) specification.

## Other information

Internal Issue (If applicable):

### CODE REVIEW

**Comment by Youssef1313** [comment]
Need some help regarding the CI failure, which shows a crash in `DrawShapedText`.


1. I can't figure out which test crashed the CI so that I can try locally.

2. The method implementation from SkiaSharp is:
 
 
 
 https://github.com/mono/SkiaSharp/blob/2ad29861d5a40d3bf78c28ab0a9cb02a8f0fe437/source/SkiaSharp.HarfBuzz/SkiaSharp.HarfBuzz.Shared/CanvasExtensions.cs#L10-L17

    It seems like `paint.GetFont().Typeface` is returning null. Is null expected in this code path? If so, is SkiaSharp somehow supposed to handle that on their side?

File: src/Directory.Build.targets
Comment (MartinZikmund) [inline src/Directory.Build.targets:48]: Maybe there is additional space here?

File: src/Uno.UWP/Uno.Skia.csproj
Comment (MartinZikmund) [inline src/Uno.UWP/Uno.Skia.csproj:43]: Also here it seems the spaces don't match

**Review by MartinZikmund — APPROVED** [review]

File: src/Directory.Build.targets
Comment (Youssef1313) [reply src/Directory.Build.targets:48]: It's Visual Studio doesn't respect spaces vs tab preference in those files I believe. Will fix.

**Comment by Youssef1313** [comment]
@MartinZikmund Can you please help on https://github.com/unoplatform/uno/pull/6785#issuecomment-899774265?

**Comment by jeromelaban** [comment]
Looks like this failed for a good reason:
```
GLIB UNHANDLED EXCEPTIONSystem.ArgumentNullException: Value cannot be null. (Parameter 'typeface')
   at SkiaSharp.HarfBuzz.SKShaper..ctor(SKTypeface typeface)
   at SkiaSharp.HarfBuzz.CanvasExtensions.DrawShapedText(SKCanvas canvas, String text, Single x, Single y, SKPaint paint)
   at Windows.UI.Composition.TextVisual.Render(SKSurface surface, SKImageInfo info) in D:\a\1\s\src\Uno.UI\UI\Xaml\Controls\TextBlock\TextVisual.skia.cs:line 159
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 86
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.RenderVisual(SKSurface surface, SKImageInfo info, Visual visual) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 98
   at Windows.UI.Composition.Compositor.Render(SKSurface surface, SKImageInfo info) in D:\a\1\s\src\Uno.UWP\UI\Composition\Compositor.skia.cs:line 52
```

**Comment by Youssef1313** [comment]
@jeromelaban Yeah, as I said in https://github.com/unoplatform/uno/pull/6785#issuecomment-899774265, I'm not sure whether a null is expected in this code path or not. I'm also not sure which test is failing to reproduce locally.

**Comment by Youssef1313** [comment]
@jeromelaban Can you review please? Thanks!

File: src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs
Comment (jeromelaban) [inline src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs:158]: Do you know in which condition this can happen? This feels like we should know more about the conditions when this can happen.

This change replaces the whole text rendering stack, and for the better. It's looking pretty good. Do you know if this could be a replacement solution for https://github.com/unoplatform/uno/pull/6975 ?

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs:158]: @jeromelaban I don't think any of them can be a replacement for the other. Both PRs solve a different problem. The other issue is that the font doesn't support the character.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs:158]: > Do you know in which condition this can happen? This feels like we should know more about the conditions when this can happen.

Not much that I know, but the crash happens in SkiaSharp at this line:

https://github.com/mono/SkiaSharp/blob/2ad29861d5a40d3bf78c28ab0a9cb02a8f0fe437/source/SkiaSharp.HarfBuzz/SkiaSharp.HarfBuzz.Shared/CanvasExtensions.cs#L10-L17

when `paint.GetFont().Typeface` returns null.

Comment (jeromelaban) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs:158]: We'll need to open an issue there then, I don't feel comfortable integrating something where there's an arbitrary try/catch.

Comment (Youssef1313) [reply src/Uno.UI/UI/Xaml/Controls/TextBlock/TextVisual.skia.cs:158]: @jeromelaban I think there could be some way to detect that beforehand. Will it be okay if we use an if condition instead of a try/catch?

**Comment by Youssef1313** [comment]
@jeromelaban Can you take a look please? Thanks!

**Comment by Youssef1313** [comment]
@jeromelaban anything left here?

**Comment by jeromelaban** [comment]
Taking a look at the CI results, here's the result for `Simple_Text`: 

image

Do you see the same thing? Are we missing a font?

**Comment by Youssef1313** [comment]
@jeromelaban Last time I ran the samples app with my changes it was working correctly. I'll re-run and see.

**Review by Youssef1313 — COMMENTED** [review]
image

@jeromelaban This is what I'm seeing locally

**Comment by Youssef1313** [comment]
@jeromelaban The font I see used locally is Segoe UI

image

**Comment by jeromelaban** [comment]
Got it. This would mean the CI is not having a compatible font (it runs on macOS).

Can you move the SimpleText RTL sample to a separate one so it's easier to troubleshoot ? Thanks!

**Comment by Youssef1313** [comment]
@jeromelaban Moved to new sample.

**Comment by Youssef1313** [comment]
@jeromelaban For some reason I cannot get a clean diff. It seems that whatever I do, the changes include line endings.

**Comment by Youssef1313** [comment]
@jeromelaban This should be ready to merge.

**Comment by jeromelaban** [comment]
Ah there's a conflict with the changes I made for OpenGL rendering in Skia, should be an easy fix.

**Comment by jeromelaban** [comment]
> @jeromelaban Is this ready now?

It is, I'm restarting the build so it validates properly.

**Comment by workgroupengineering** [comment]
@Youssef1313 
Have you seen this https://github.com/woutest79/SkiaSharp.TextBlocks ?. It supports multiple scenarios.
I was planning to create a PR for UNO , but right now I don't have the time.

**Comment by Youssef1313** [comment]
I haven't checked https://github.com/wouterst79/SkiaSharp.TextBlocks. What are additional benefits it can provide more than the current approach?

Tagging @jeromelaban for whether he is okay to add this dependency if it can provide more benefits.

**Comment by workgroupengineering** [comment]
> 



> I haven't checked https://github.com/wouterst79/SkiaSharp.TextBlocks. What are additional benefits it can provide more than the current approach?
> 
> Tagging @jeromelaban for whether he is okay to add this dependency if it can provide more benefits.
> 
> >  GitHub**[GitHub - wouterst79/SkiaSharp.TextBlocks: SkiaSharp.TextBlock adds text block and rich text layout to SkiaSharp](https://github.com/wouterst79/SkiaSharp.TextBlocks)**SkiaSharp.TextBlock adds text block and rich text layout to SkiaSharp  - GitHub - wouterst79/SkiaSharp.TextBlocks: SkiaSharp.TextBlock adds text block and rich text layout to SkiaSharp

- Text alignment 
- Unicode symbols
- line spacing

**Comment by Youssef1313** [comment]
@workgroupengineering I'm going to open a separate issue for this.

**Comment by francoistanguay** [comment]
We have a large refactoring planned on Skia Text. I wouldnt invest any time further on this until we do it in the upcoming days/weeks.

**Comment by Youssef1313** [comment]
@francoistanguay Please let me know when this happens. Thanks!

**Comment by MartinZikmund** [comment]
This might now be taken care of by @mikernet PR?

**Comment by Youssef1313** [comment]
@MartinZikmund I don't know if it's. We can confirm after it's merged.

**Comment by jeromelaban** [comment]
The PR has been merged, let's revalidate the original issue.

**Comment by MartinZikmund** [comment]
@Youssef1313 Seems the current font rendering got just halfway there - screenshot from the issue repo:
image

**Comment by mikernet** [comment]
I can PR a fix for RTL rendering today, it's just a segment ordering issue. Bi-directional text won't be supported yet but text that is all RTL will work correctly.

**Comment by Youssef1313** [comment]
@mikernet Please do that! Meanwhile, I'm going to close this PR.

**Comment by mikernet** [comment]
@Youssef1313 I should probably hold off on PRing this because I'm stacking PRs on top of other PRs that haven't been merged yet, it's going to get messy if I keep going, lol. I'll do this after https://github.com/unoplatform/uno/pull/9285 is merged.

**Pull request closed without merging**
